### ISSUE TITLE
Feature needed for HIPAA Compliance 

### ISSUE THREAD

**Issue by oren** [convox/kernel#92]
* Dedicated EC2
* Private subnet for your applications
* Public subnet for NAT and VPN
* Jump Box (Bastion) in the public subnet with logging
* Encryption even inside the private subnet (for example between the app to the DB)

[AWS Whitepaper on HIPAA](http://d0.awsstatic.com/whitepapers/compliance/AWS_HIPAA_Compliance_Whitepaper.pdf)

**Comment by oren** [comment]
@ddollar's reply:

I think the public and private thing won't be too difficult. Depends on how private it needs to be
If it's only reachable by load balancers and not from the public internet is that private enough?
So no internet routable ip on the boxes.
Internal encryption will happen at the app later by using TLS to communicate with everything. 
The bastion with logging will probably be the most time consuming part but if the aforementioned network configuration with load balancers is acceptable we may be able to make an ELB that you can SSH through that has logging.

**Comment by ddollar** [comment]
I think the public and private thing won't be too difficult. Depends on how private it needs to be
If it's only reachable by load balancers and not from the public internet is that private enough?

The boxes would have no internet-routable IP assumed. 

Internal encryption will happen at the app layer by using TLS to communicate with everything. 

The bastion with logging will probably be the most time consuming part but if the aforementioned network configuration with load balancers is acceptable we may be able to make an ELB that you can SSH through that has logging.

It should also be fairly simple to set up a VPN gateway from your Convox VPC to wherever you need

**Comment by oren** [comment]
More information about 'The ability to deploy a container to specific EC2(s)'.

Here are two cases:
1. VPN software should only run on the EC2 that is on the public subnet.
2. A customer might ask to have all it's apps running alone, meaning that no other containers are allowed to run on a server that running this customer's apps.

**Comment by nzoschke** [comment]
@oren we got some more information about ECS and HIPAA:

> ECS has not yet gone through the HIPAA certification process...
> ECS is not processing HIPAA information and the only thing that might be sensitive could be environment variables but that will be up to [your] assessor to decide.

**Comment by stevegraham** [comment]
Even if it's not HIPAA eligible, a lot of other classes of application would benefit from similar features :)

**Comment by oren** [comment]
@stevegraham Agree. HIPAA forces you to level-up the maturity of your infrastructure and processes and even if you don't need to be HIPAA compliance, it's a good goal so strive for.

**Comment by ddollar** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;h0QCyHBm&#x2F;122-hipaa-compliance-checklist"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> HIPAA compliance checklist</a>

**Comment by seanknox** [comment]
(disclaimer: I work at a healthcare company subject to HIPAA that stores PHI on AWS, but I'm not a lawyer/assessor). 

Hi, just want to note that ECS can be used in orgs building toward HIPAA compliance. tldr; ECS doesn't itself handle PHI, it orchestrates EC2 servers that may have PHI stored. EC2 *is* a HIPAA-eligible service.

From our Amazon SE:

> The EC2 Container Service (ECS) is simply orchestrating and scheduling application containers which are deployed and run on cluster in EC2; the actual PHI is processed on EC2 which is an Eligible Service under the BAA. Customers must still ensure that EC2 instances processing, storing or transmitting PHI be launched in Dedicated Tenancy and that PHI is encrypted at rest and in transit. Any application meta-data stored as a TaskDefinition or in ECS may not contain PHI, so the definition of what applications to start should not contain any PHI. Additionally, the exit string of a failed container should not contain PHI as this data is persisted in the ECS service itself, outside of EC2.

You still need to be careful about logging, but as long as you follow the rest of the requirements you should be fine.

While I'm here, how does one deploy an EC2 instance with dedicated tenancy using Convox? I haven't found that bit yet.

**Comment by oren** [comment]
@seanknox very insightful. thank you! 
I believe that what you are looking for is `convox install --dedicated`
